contracts/upgradeable_contracts/erc20_to_erc20/HomeBridgeErcToErc.sol

```diff
@@ -164,13 +188,19 @@ contract HomeBridgeErcToErc is
         }
     }
 
-    function onFailedAffirmation(address _recipient, uint256 _value, bytes32 _txHash) internal {
-        address recipient;
-        uint256 value;
-        (recipient, value) = txAboveLimits(_txHash);
+    /**
+     * @dev Internal callback to be called on failed message execution due to the out-of-limits error.
+     * This function saves the bridge operation information for further processing.
+     * @param _recipient address of the receiver where the new tokens should be minted.
+     * @param _value amount of tokens to mint.
+     * @param _txHash reference transaction hash on the Foreign side of the bridge which cause this operation.
+     * @param _hashMsg unique identifier of the particular bridge operation.
+     */
+    function onFailedAffirmation(address _recipient, uint256 _value, bytes32 _txHash, bytes32 _hashMsg) internal {
+        (address recipient, uint256 value) = txAboveLimits(_hashMsg);
         require(recipient == address(0) && value == 0);
         setOutOfLimitAmount(outOfLimitAmount().add(_value));
-        setTxAboveLimits(_recipient, _value, _txHash);
-        emit AmountLimitExceeded(_recipient, _value, _txHash);
+        setTxAboveLimits(_recipient, _value, _hashMsg);
+        emit AmountLimitExceeded(_recipient, _value, _txHash, _hashMsg);
```

Comment:
Do we need the tx hash here at all? In which case it could be necessary? Will it possible to extract it from the transaction input data in such case?

Reply:
I thought it can be useful in a situation when we want to find out if a given transfer was marked as out of limits.
In current implementation, we can search logs by using `txHash` directly.
But if we remove the `txHash` out from this event, then we need to construct a message manually, then take its hash, and search through the logs with this particular message hash.

It is not a big problem, so it is possible to remove it and keep the only version of `AmountLimitExceeded` event.
What do you think?

Reply:
Ok, it makes sense to leave both the tx hash and the msg hash both in the event to allow easily to look up if an affirmation request was failed or not. Then consider to have messageId (for the event with 3 arguments) and txHash (for the event with 4 arguments) to be indexed in the event description in order to simplify the process.

Probably, it makes sense to provide different name for these events like AmountLimitExceededInAMBExt and AmountLimitExceeded, doesn't it?

Consider